examples/interpreter-demo/serialize.js

```diff
@@ -0,0 +1,411 @@
+        break;
+      case 'State':
+        obj = new Interpreter.State(undefined, undefined);
+        break;
+      case 'Node':
+        obj = new NODE_CONSTRUCTOR();
+        delete obj.start;
+        delete obj.end;
+        var locText = jsonObj['loc'];
+        if (locText) {
+          var loc = new NODE_LOC_CONSTRUCTOR();
+          var m = locText.match(LOC_REGEX);
+          var locStart = null;
+          if (m[1] || m[2]) {
+            locStart = new LINE_LOC_CONSTRUCTOR();
+            locStart.line = Number(m[1]);
+            locStart.column = Number(m[2]);
+          }
+          loc.start = locStart;
+          var locEnd = null;
+          if (m[3] || m[4]) {
+            locEnd = new LINE_LOC_CONSTRUCTOR();
+            locEnd.line = Number(m[3]);
+            locEnd.column = Number(m[4]);
+          }
+          loc.end = locEnd;
+          if (m[5]) {
+            loc.source = decodeURI(m[5]);
+          } else {
+            delete loc.source;
+          }
+          obj.loc = loc;
+        }
+        break;
+      default:
+        throw TypeError('Unknown type: ' + jsonObj['type']);
+    }
+    objectList[i] = obj;
+  }
+  // Second pass: Populate properties for every object.
+  for (var i = 0; i < json.length; i++) {
+    var jsonObj = json[i];
+    var obj = objectList[i];
+    // Repopulate objects.
+    var props = jsonObj['props'];
+    if (props) {
+      var nonConfigurable = jsonObj['nonConfigurable'] || [];
+      var nonEnumerable = jsonObj['nonEnumerable'] || [];
+      var nonWritable = jsonObj['nonWritable'] || [];
+      var getter = jsonObj['getter'] || [];
+      var setter = jsonObj['setter'] || [];
+      var names = Object.getOwnPropertyNames(props);
+      for (var j = 0; j < names.length; j++) {
+        var name = names[j];
+        var descriptor = {
+          configurable: nonConfigurable.indexOf(name) === -1,
+          enumerable: nonEnumerable.indexOf(name) === -1
+        };
+        var hasGetter = getter.indexOf(name) !== -1;
+        var hasSetter = setter.indexOf(name) !== -1;
+        if (hasGetter || hasSetter) {
+          if (hasGetter) {
+            descriptor.get = interpreter.setProperty.placeholderGet_;
+          }
+          if (hasSetter) {
+            descriptor.set = interpreter.setProperty.placeholderSet_;
+          }
+        } else {
+          descriptor.value = decodeValue(props[name]);
+          descriptor.writable = nonWritable.indexOf(name) === -1;
+        }
+        Object.defineProperty(obj, name, descriptor);
+      }
+    }
+    // Repopulate arrays.
+    if (Array.isArray(obj)) {
+      var data = jsonObj['data'];
+      if (data) {
+        for (var j = 0; j < data.length; j++) {
+          obj.push(decodeValue(data[j]));
+        }
+      }
+    }
+  }
+  // First object is the interpreter.
+  var root = objectList[0];
+  for (var prop in root) {
+    interpreter[prop] = root[prop];
+  }
+}
+
+function serialize(interpreter) {
```

Comment:
Also decompose this function.

Reply:
This is the focus-point of the design.  The offending `encodeValue` function needs access to `objectList`.  And one of the two calls is from `Array.map`.  Thus there's no way to pass `objectList` in as a second argument.  The options are:

1. Wrap the call as a new function: `map(function(element) {return encodeValue(element, objectList);})` This creates a new function for every array in the interpreter, which is a lot.
2. Reverse the order of the arguments and bind the function: `map(encodeValue.bind(null, objectList))` This creates a new function for every array in the interpreter, and also would be much harder to understand.
3. Push `objectList` into a global variable.  Works fine.
4. Push `encodeValue` into `serialize`.  Also works fine.

The fourth option is what we currently have.  I that's preferable to a global variable, let alone the other options.  What do you think?

Reply:
I see. 3 is fine then, particularly since (as previously discussed) this code is not new code written for this particular demo.

At a higher level, this is what modules are for: letting you put things in file-global variables without having them be accessible everywhere.

Reply:
Done.